src/Common/AsyncLoader.cpp

```diff
@@ -0,0 +1,607 @@
+    NOEXCEPT_SCOPE({
+        ALLOW_ALLOCATIONS_IN_SCOPE;
+        finished_jobs.insert(job);
+        logAboutProgress(log, finished_jobs.size() - old_jobs, finished_jobs.size() + scheduled_jobs.size() - old_jobs, stopwatch);
+    });
+}
+
+void AsyncLoader::prioritize(const LoadJobPtr & job, ssize_t new_priority, std::unique_lock<std::mutex> & lock)
+{
+    if (auto info = scheduled_jobs.find(job); info != scheduled_jobs.end())
+    {
+        if (info->second.priority >= new_priority)
+            return; // Never lower priority
+
+        // Update priority and push job forward through ready queue if needed
+        if (info->second.ready_seqno)
+            ready_queue.erase(info->second.key());
+        info->second.priority = new_priority;
+        job->load_priority.store(new_priority); // Set user-facing priority (may affect executing jobs)
+        if (info->second.ready_seqno)
+        {
+            NOEXCEPT_SCOPE({
+                ALLOW_ALLOCATIONS_IN_SCOPE;
+                ready_queue.emplace(info->second.key(), job);
+            });
+        }
+
+        // Recurse into dependencies
+        for (const auto & dep : job->dependencies)
+            prioritize(dep, new_priority, lock);
+    }
+}
+
+void AsyncLoader::enqueue(Info & info, const LoadJobPtr & job, std::unique_lock<std::mutex> & lock)
+{
+    chassert(!info.is_blocked());
+    chassert(info.ready_seqno == 0);
+    info.ready_seqno = ++last_ready_seqno;
+    NOEXCEPT_SCOPE({
+        ALLOW_ALLOCATIONS_IN_SCOPE;
+        ready_queue.emplace(info.key(), job);
+    });
+
+    if (is_running && workers < max_threads)
+        spawn(lock);
+}
+
+void AsyncLoader::spawn(std::unique_lock<std::mutex> &)
+{
+    workers++;
+    NOEXCEPT_SCOPE({
+        ALLOW_ALLOCATIONS_IN_SCOPE;
+        pool.scheduleOrThrowOnError([this] { worker(); });
+    });
+}
+
+void AsyncLoader::worker()
+{
+    DENY_ALLOCATIONS_IN_SCOPE;
+
+    LoadJobPtr job;
+    std::exception_ptr exception_from_job;
+    while (true)
+    {
+        // This is inside the loop to also reset previous thread names set inside the jobs
+        setThreadName("AsyncLoader");
+
+        {
+            std::unique_lock lock{mutex};
+
+            // Handle just executed job
+            if (exception_from_job)
+                finish(lock, job, LoadStatus::FAILED, exception_from_job);
+            else if (job)
+                finish(lock, job, LoadStatus::OK);
+
+            if (!is_running || ready_queue.empty() || workers > max_threads)
+            {
+                workers--;
+                return;
+            }
+
+            // Take next job to be executed from the ready queue
+            auto it = ready_queue.begin();
+            job = it->second;
+            ready_queue.erase(it);
+            scheduled_jobs.find(job)->second.ready_seqno = 0; // This job is no longer in the ready queue
+        }
+
+        ALLOW_ALLOCATIONS_IN_SCOPE;
+
+        try
+        {
+            job->execute(job);
```

Comment:
this looks strange, why can't we send a raw pointer inside `execute` with `this`?

Reply:
`LoadJobPtr` can be useful because currently there is no way to get a smart pointer from raw `LoadJob`. And smart pointer is required in every call in AsyncLoader API. If you are going to create more jobs/tasks that should be run no earlier than "this-job-finish" it might be useful. Though, we can use `std::enable_shared_from_this`, I find it too complicated, so I usually try to avoid it if possible.

Reply:
yes, let's avoid doing `std::enable_shared_from_this`